File: 2024/8/solution.js
const fs = require("fs");
const path = require("path");

const input = fs
  .readFileSync(path.join(__dirname, "input.txt"), "utf-8")
  .split("\n");

const antennaMap = new Map();
for (let r = 0; r < input.length; r++) {
  for (let c = 0; c < input[0].length; c++) {
    if (input[r][c] !== ".") {
      if (!antennaMap.has(input[r][c])) {
        antennaMap.set(input[r][c], []);
      }
      antennaMap.get(input[r][c]).push([r, c]);
    }
  }
}

const hgt = input.length;
const wid = input[0].length;

const antinodes = new Set();
const antinodes2 = new Set();

for (const coords of antennaMap.values()) {
  for (let i = 0; i < coords.length; i++) {
    for (let j = i + 1; j < coords.length; j++) {
      const [r1, c1] = coords[i];
      const [r2, c2] = coords[j];
      const dr = r2 - r1;
      const dc = c2 - c1;
      if (r1 - dr >= 0 && r1 - dr < hgt && c1 - dc >= 0 && c1 - dc < wid)
        antinodes.add(`${r1 - dr},${c1 - dc}`);
      let k = 0;
      while (
        r1 - dr * k >= 0 &&
        r1 - dr * k < hgt &&
        c1 - dc * k >= 0 &&
        c1 - dc * k < wid
      ) {
        antinodes2.add(`${r1 - dr * k},${c1 - dc * k}`);
        k++;
      }
      if (r2 + dr >= 0 && r2 + dr < hgt && c2 + dc >= 0 && c2 + dc < wid)
        antinodes.add(`${r2 + dr},${c2 + dc}`);
      k = 0;
      while (
        r2 + dr * k >= 0 &&
        r2 + dr * k < hgt &&
        c2 + dc * k >= 0 &&
        c2 + dc * k < wid
      ) {
        antinodes2.add(`${r2 + dr * k},${c2 + dc * k}`);
        k++;
      }
    }
  }
}

console.log(antinodes.size);
console.log(antinodes2.size);

File: 2024/8/solution.go
package main

import (
	"fmt"
	"os"
	"path/filepath"
	"runtime"
	"strings"
)

func main() {
	_, filename, _, _ := runtime.Caller(0)
	dirname := filepath.Dir(filename)
	inputFilePath := filepath.Join(dirname, "input.txt")
	data, err := os.ReadFile(inputFilePath)
	if err != nil {
		panic(err)
	}
	puzzleInput := strings.Split(string(data), "\n")
	fmt.Println(part1(puzzleInput))
}

func part1(puzzleInput []string) int {
	antinodeMap := make(map[[2]int]struct{})
	antennaMap := make(map[byte][][2]int)
	rMax := len(puzzleInput) - 1
	cMax := len(puzzleInput[0]) - 1
	for r := range puzzleInput {
		for c := range puzzleInput[r] {
			if puzzleInput[r][c] != '.' {
				antennaMap[puzzleInput[r][c]] = append(antennaMap[puzzleInput[r][c]], [2]int{r, c})
			}
		}
	}
	for _, pairs := range antennaMap {
		for i, antenna1 := range pairs {
			for _, antenna2 := range pairs[i+1:] {
				dr := antenna2[0] - antenna1[0]
				dc := antenna2[1] - antenna1[1]
				r1, c1 := antenna1[0]-dr, antenna1[1]-dc
				r2, c2 := antenna2[0]+dr, antenna2[1]+dc
				if r1 >= 0 && c1 >= 0 && r1 <= rMax && c1 <= cMax {
					antinodeMap[[2]int{r1, c1}] = struct{}{}
				}
				if r2 >= 0 && c2 >= 0 && r2 <= rMax && c2 <= cMax {
					antinodeMap[[2]int{r2, c2}] = struct{}{}
				}
			}
		}
	}
	return len(antinodeMap)
}
